unstructured_ingest/v2/processes/connectors/discord.py

```diff
@@ -0,0 +1,149 @@
+import datetime as dt
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Any, Generator, Optional
+
+from pydantic import Field, Secret
+
+from unstructured_ingest.utils.dep_check import requires_dependencies
+from unstructured_ingest.v2.interfaces import (
+    AccessConfig,
+    ConnectionConfig,
+    Downloader,
+    DownloaderConfig,
+    DownloadResponse,
+    FileData,
+    FileDataSourceMetadata,
+    Indexer,
+    IndexerConfig,
+    SourceIdentifiers,
+)
+from unstructured_ingest.v2.logger import logger
+from unstructured_ingest.v2.processes.connector_registry import SourceRegistryEntry
+
+if TYPE_CHECKING:
+    from discord import Client as DiscordClient
+
+CONNECTOR_TYPE = "discord"
+
+
+class DiscordAccessConfig(AccessConfig):
+    token: str = Field(description="Discord API token")
+
+
+class DiscordConnectionConfig(ConnectionConfig):
+    access_config: Secret[DiscordAccessConfig] = Field(
+        default=DiscordAccessConfig, validate_default=True
+    )
+
+    @requires_dependencies(["discord"], extras="discord")
+    def get_client(self) -> "DiscordClient":
+        import discord
+
+        intents = discord.Intents.default()
+        intents.message_content = True
+        return discord.Client(intents=intents)
+
+
+class DiscordIndexerConfig(IndexerConfig):
+    channels: Optional[list[str]] = Field(
+        default=None, description="List of Discord channel IDs to process"
+    )
+
+
+@dataclass
+class DiscordIndexer(Indexer):
+    connection_config: DiscordConnectionConfig
+    index_config: DiscordIndexerConfig
+
+    def run(self, **kwargs: Any) -> Generator[FileData, None, None]:
+        self.connection_config.get_client()
+        channels_to_process: set[str] = set(self.index_config.channels or [])
+
+        for channel_id in list(channels_to_process):
+            file_data = self.get_channel_file_data(channel_id=channel_id)
+            if file_data:
+                yield file_data
+
+    def get_channel_file_data(self, channel_id: str) -> Optional[FileData]:
+        # Fetch channel metadata
+        identifier = channel_id
+        channel_id = f"{channel_id}.txt"
+        source_identifiers = SourceIdentifiers(
+            filename=channel_id,
+            fullpath=channel_id,
+        )
+        metadata = FileDataSourceMetadata(
+            record_locator={"channel_id": identifier},
+            date_processed=str(dt.datetime.utcnow().isoformat()),
+        )
+        return FileData(
+            identifier=identifier,
+            connector_type=CONNECTOR_TYPE,
+            source_identifiers=source_identifiers,
+            metadata=metadata,
+        )
+
+
+class DiscordDownloaderConfig(DownloaderConfig):
+    limit: Optional[int] = Field(
```

Comment:
Do we want to support a time order on this? So if limit is applied, pull in the latest messages, or the oldest? If this isn't easy to do, I won't worry about it, just a thought.

Reply:
Right, so by default it takes the latest. Fortunately the method used has a sensible API to control it:
```
    async def history(
        self,
        *,
        limit: Optional[int] = 100,
        before: Optional[SnowflakeTime] = None,
        after: Optional[SnowflakeTime] = None,
        around: Optional[SnowflakeTime] = None,
        oldest_first: Optional[bool] = None,
    ) -> AsyncIterator[Message]:
    ```